File: dijkstra.js
function dijkstra(graph, start) {
    const distances = {};
    const visited = new Set();
    
    // Initialize all distances to infinity
    for (let node in graph) {
        distances[node] = Infinity;
    }
    distances[start] = 0;
    
    // Priority queue implementation using array (for simplicity)
    const priorityQueue = [{distance: 0, node: start}];
    
    while (priorityQueue.length > 0) {
        // Sort by distance to get minimum (simple priority queue)
        priorityQueue.sort((a, b) => a.distance - b.distance);
        const {distance: currentDistance, node: current} = priorityQueue.shift();
        
        // Skip if already visited
        if (visited.has(current)) {
            continue;
        }
        visited.add(current);
        
        // Skip if we've already found a shorter path
        if (currentDistance > distances[current]) {
            continue;
        }
        
        // Check all neighbors
        if (graph[current]) {
            for (let neighbor of graph[current]) {
                const neighborNode = neighbor.node;
                const weight = neighbor.weight;
                const newDistance = distances[current] + weight;
                
                // If we found a shorter path, update it
                if (newDistance < distances[neighborNode]) {
                    distances[neighborNode] = newDistance;
                    priorityQueue.push({distance: newDistance, node: neighborNode});
                }
            }
        }
    }
    
    return distances;
}

File: dijkstra.cpp
#include <iostream>
#include <vector>
#include <unordered_map>
#include <queue>
#include <climits>
using namespace std;

unordered_map<int, int> dijkstra(unordered_map<int, vector<pair<int, int>>>& graph, int start) {
    unordered_map<int, int> distances;
    
    // Initialize all distances to infinity
    for (auto& node : graph) {
        distances[node.first] = INT_MAX;
    }
    distances[start] = 0;
    
    // Priority queue to store (distance, node) pairs
    priority_queue<pair<int, int>, vector<pair<int, int>>, greater<pair<int, int>>> pq;
    pq.push({0, start});
    
    while (!pq.empty()) {
        int current_distance = pq.top().first;
        int current = pq.top().second;
        pq.pop();
        
        // Skip if we've already found a shorter path
        if (current_distance > distances[current]) {
            continue;
        }
        
        // Check all neighbors
        for (auto& neighbor_pair : graph[current]) {
            int neighbor = neighbor_pair.first;
            int weight = neighbor_pair.second;
            int new_distance = distances[current] + weight;
            
            // If we found a shorter path, update it
            if (new_distance < distances[neighbor]) {
                distances[neighbor] = new_distance;
                pq.push({new_distance, neighbor});
            }
        }
    }
    
    return distances;
}
